contents/java/Java.md

```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+The event bus uses reflections to observe all registered objects via `register` method for methods annotated with the
+`Subscribe` annotation.
+
+An example implementation (not the actual) of the `Subscribe` annotation might be:
+
+```java
+@Retention(RetentionPolicy.RUNTIME) // Retain this annotation at runtime!
+@Target(ElementType.METHOD) // Only can be applied to methods!
+public @interface Subscribe { }
+```
+
+And that is it! Important parts of the code to note is the first and second line before the declaration. The first line
+tells Java that this annotation must not be discarded during compile time so it will be available during runtime. The
+retention poilcy is there because some annotation do not mean anything after compilation (such as `Override` and
+`SuppressWarnings`), so it does not make sense to keep it after compiling.
+
+The second line just means that this annotation can be applied to methods only.
+
+And the more important part is the Subscriber Registry
+
+```java
+// TypeToken class is provided by Google Guava reflection library
+Set<? extends Class<?>> supertypes = TypeToken.of(clazz).getTypes().rawTypes();
+
+for (Class<?> supertype : supertypes) {
+      for (Method method : supertype.getDeclaredMethods()) {
+        if (method.isAnnotationPresent(Subscribe.class) {
+            registerSubscriber(method);
+        }
+    }
+}
+```
+
+The first line finds all classes and its parent classes of the registered class and converts it to a set. That is if you
+registered `Sheep extends Animal` as an event handler, both `Sheep` and `Animal` will be captured by the first line. The
+following lines will then examine all their methods (during runtime!) for the subscribe annotation and register it so
+that it will receive the specified event when it is fired. Of course this implementation leaves out a lot of details but
+you get the idea of how java reflections
+
+### Disadvantages of reflections
+
+While Java reflection is powerful, you should not immediately jump on the reflections ship. This is because there are
+some drawbacks whenever you use reflections in your project.
+
+* Converts a compile-time error to a potentially destructive run-time error.
+
+Compile time errors are easy to catch. Whenever you compile your code, Java's compiler cleverly spots any error you
+missed and points it out (along with line number and other useful information) to you before quitting. But by using
+reflections, you are bypassing these checks because there is no way to check such errors during compile time.
+Runtime errors are more troublesome in that they are harder to catch and debug and might even bring your whole software
+under the water with it.
+
+* Reflections are harder to understand and harder to debug
+
+There is a reason why reflections is placed under the advanced section. Codes using reflections are fundamentally
```

Comment:
"...the topic of reflections is..." instead of "reflections is"? (seems like a subject-verb agreement mistake)